### PR TITLE
Port DocumentState and UnknownDocument.

### PR DESCRIPTION
**Pull request by rsgowman**
Part of heldwriteacks.

Serialization work for this is largely deferred until after
nanopb-master is merged with master.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/model/document.h [hunk 1]
```diff
@@ -26,6 +26,21 @@ namespace firebase {
 namespace firestore {
 namespace model {
 
+enum class DocumentState {
+  /**
+   * Local mutations applied via the mutation queue. Document is potentially
+   * inconsistent.
+   */
+  kLocalMutations,
+  /**
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/document.h]: Optional nit: maybe add a new line to visually separate the elements?

File: Firestore/core/src/firebase/firestore/model/document.h [hunk 2]
```diff
@@ -49,15 +64,23 @@ class Document : public MaybeDocument {
...
   bool has_local_mutations() const {
-    return has_local_mutations_;
+    return document_state_ == DocumentState::kLocalMutations;
+  }
+
+  bool has_committed_mutations() const {
+    return document_state_ == DocumentState::kCommittedMutations;
+  }
+
+  bool HasPendingWrites() const override {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/document.h]: Question: this function uses PascalCase, while the other two very similar functions use snake_case. Can these be in the same style? (no strong preference which)

File: Firestore/core/src/firebase/firestore/model/mutation.cc [hunk 3]
```diff
@@ -84,17 +84,13 @@ std::shared_ptr<const MaybeDocument> PatchMutation::ApplyToLocalView(
   VerifyKeyMatches(maybe_doc.get());
 
   if (!precondition().IsValidFor(maybe_doc.get())) {
-    if (maybe_doc) {
-      return absl::make_unique<MaybeDocument>(maybe_doc->key(),
-                                              maybe_doc->version());
-    }
-    return nullptr;
+    return maybe_doc;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/mutation.cc]: What's the reason for this change?

File: Firestore/core/src/firebase/firestore/model/no_document.cc [hunk 4]
```diff
@@ -22,8 +22,11 @@ namespace firebase {
 namespace firestore {
 namespace model {
 
-NoDocument::NoDocument(DocumentKey key, SnapshotVersion version)
-    : MaybeDocument(std::move(key), std::move(version)) {
+NoDocument::NoDocument(DocumentKey key,
+                       SnapshotVersion version,
+                       bool has_committed_mutations)
+    : MaybeDocument(std::move(key), std::move(version)),
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/no_document.cc]: Optional nit: moving `SnapshotVersion` is no more efficient than copying it (unless there's an expectation this class will grow) (here and in some other constructors)

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 5]
```diff
@@ -497,8 +498,9 @@ std::unique_ptr<MaybeDocument> Serializer::DecodeBatchGetDocumentsResponse(
   } else if (found != nullptr) {
     return found;
   } else if (!missing.empty()) {
-    return absl::make_unique<NoDocument>(
-        DecodeKey(missing), SnapshotVersion{*std::move(read_time)});
+    return absl::make_unique<NoDocument>(DecodeKey(missing),
+                                         SnapshotVersion{*std::move(read_time)},
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: Same optional nit about moving.

File: Firestore/core/src/firebase/firestore/model/maybe_document.h [hunk 6]
```diff
@@ -41,6 +41,7 @@ class MaybeDocument {
     Unknown,
     Document,
     NoDocument,
+    UnknownDocument,
```
Comment (zxu123) [inline Firestore/core/src/firebase/firestore/model/maybe_document.h]: Since you are adding type here, could you add document for each of the type? In particular, it helps reader differing `Unknown` and `UnknownDocument`.

File: Firestore/core/test/firebase/firestore/model/CMakeLists.txt [hunk 7]
```diff
@@ -21,7 +21,6 @@ cc_test(
     field_mask_test.cc
     field_path_test.cc
     field_value_test.cc
-    maybe_document_test.cc
```
Comment (zxu123) [inline Firestore/core/test/firebase/firestore/model/CMakeLists.txt]: Why remove this unit test?

File: Firestore/core/src/firebase/firestore/model/no_document.cc (hunk 4 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/model/no_document.cc]: For non-primitive types, I usually just std::move it, unless it's *really* obvious that the thing is small. (really obvious => documented as small.) We can probably assume that with SnapshotVersion (and Timestamp), since it seems unlikely that either will grow.

Somewhat related: http://go/c-readability-advice#value-param

Proposal: Since you've called this optional, and since we std::move SnapshotVersion's (and probably Timestamp's too) somewhat inconsistently in the rest of the code, I'll defer this for now. But I'll create a followup inspired by the above readability link that:
a) ensures std::is_trivially_copyable()
b) ensures sizeof <= 16
for these types, and then:
c) remove usages of std::move and const&.

File: Firestore/core/src/firebase/firestore/model/maybe_document.h (hunk 6 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/model/maybe_document.h]: I'd prefer not to add documentation for each type, since that can already be found in the subclasses themselves. So instead, I've added "See the docstrings of the subclasses for details." to the docstring of the enum itself.

However, "Unknown" isn't a subclass. It used to be used by MaybeDocument in situations where a MaybeDocument was created directly. That's no longer possible since MaybeDocument is now abstract. It's also used as a placeholder for an invalid value that indicates a pretty serious problem with serialization. I've added a docstring for this value.

I've also added a TODO to investigate removing it entirely; I suspect the serializers can do without it.

File: Firestore/core/test/firebase/firestore/model/CMakeLists.txt (hunk 7 above)
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/model/CMakeLists.txt]: Because MaybeDocument is now abstract, so you can't test it directly. Fortunately (?) all the tests were already duplicated in the subclass tests.
